magento/magento2
# Fix loss of page_cache cache_dir setting from di.xml

### Description

In a native Magento 2 installation both the page_cache and the default cache is stored in the `var/cache` directory. The `var/page_cache` directory is not used, even, though it is [configured in di.xml](https://github.com/magento/magento2/blob/2.3-develop/app/etc/di.xml#L771).

This makes it impossible to use `bin/magento cache:flush full_page` to only clear the FPC, as it will also flush the cache directory.

The reason is any cache options configured in `di.xml` are lost if any configuration for a cache frontend is present in `app/etc/env.php`.

During the Magento installation a cache id_prefix for both the [default](\Magento\Setup\Model\ConfigOptionsList\Cache::createConfig) and the [page_cache](\Magento\Setup\Model\ConfigOptionsList\PageCache) frontends are written to the `app/etc/env.php`.

The bug was introduced by the PR: https://github.com/magento/magento2/pull/18641  
(note: the issue that was solved by that PR is also relevant for non-filesystem cache backends.) 

In the method body of `\Magento\Framework\App\Cache\Frontend\Pool::_getCacheSettings` the original comment read:

```
Merging is intentionally implemented through array_merge() instead of array_replace_recursive()
to avoid "inheritance" of the default settings that become irrelevant as soon as cache storage type changes
```

This reasoning of the original comment applies only if the backend type is changed, but not if an option is added to be used with the default.

### Fixed Issues

This PR will merge the deployment configuration and the `di.xml` configuration using `array_replace_recursive` instead of `array_merge`, as cache settings that are not used by other cache backends don't do any harm, and still can be changed by overwriting them when needed. 

An alternative implementation of the solution checked if the `page_cache` backend was set to a different value of the `default` cache frontend, but that added a lot of additional coupling and in favor of simplicity I believe the solution in this PR is better.

The above fix revealed a bug that was masked before, namely that the when the built in page cache was is used, moving a category does not clean the appropriate cache records.
With varnish this issue does not occur because there the category top navigation is rendered as an ESI block, which is associated with all category ID cache tags (`cat_c_<ID>`).  
The page cache records however are only associated with the generic category cache tag (`cat_c`).  
Moving a category only cleans records that are associated with the category ID cache tag.  
The solution is implemented as an event observer for the `category_move` event, which cleans the generic category cache tag if the built in page cache is used.

### Manual testing scenarios

1. download source Magento (tarball or composer or github clone)
2. install Magento (ui or cli)
3. access a frontend or adminhtml page in the browser
4. With the bug: `var/page_cache` is empty and `var/cache` is not empty
5. Run `bin/magento cache:flush full_page`
6. Check `var/cache` is empty

After the patch is applied `var/page_cache` will no longer be empty after a page is accessed and `bin/magento cache:flush full_page` will only flush `var/page_cache`

### Contribution checklist
